src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,576 @@
+                        {
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                                new PipeException(SR.PipeCantCloseWithPendingWrite), ExceptionEventType);
+                        }
+
+                        if (_closeState == CloseState.Closing || _closeState == CloseState.HandleClosed)
+                        {
+                            // already closing or closed, so just return
+                            return;
+                        }
+
+                        _closeState = CloseState.Closing;
+
+                        shouldCloseHandle = true;
+
+                        if (!_isAtEOF)
+                        {
+                            if (_inReadingState)
+                            {
+                                existingReadIsPending = true;
+                            }
+                            else
+                            {
+                                shouldReadEOF = true;
+                            }
+                        }
+
+                        if (!_isShutdownWritten)
+                        {
+                            shouldWriteEOF = true;
+                            _isShutdownWritten = true;
+                        }
+                    }
+                }
+
+                ValueTask writeValueTask = default;
+                ValueTask<int> readValueTask = default;
+                if (shouldWriteEOF)
+                {
+                    writeValueTask = StartWriteZeroAsync(cancellationToken);
+                }
+
+                if (shouldReadEOF)
+                {
+                    readValueTask = StartReadZeroAsync(cancellationToken);
+                }
+
+                // wait for shutdown write to complete
+                if (shouldWriteEOF)
+                {
+                    try
+                    {
+                        await WaitForWriteZero(writeValueTask, timeout, true);
+                    }
+                    catch (TimeoutException e)
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownWriteError, e), ExceptionEventType);
+                    }
+                }
+
+                // ensure we have received EOF signal
+                if (shouldReadEOF)
+                {
+                    try
+                    {
+                        await WaitForReadZero(readValueTask, timeout, true);
+                    }
+                    catch (TimeoutException e)
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownReadError, e), ExceptionEventType);
+                    }
+                }
+                else if (existingReadIsPending)
+                {
+                    if (!TimeoutHelper.Wait(_atEOFEvent, timeoutHelper.RemainingTime()))
```

Comment:
Should EOFEvent ManualResetEventSlim be a TaskCompletionSource? It looks like this will block. A TCS would allow an await here.